src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractFormatEngine.cs

```diff
@@ -132,75 +131,41 @@ protected virtual NodeOperations CreateNodeOperations(CancellationToken cancella
+            List<AlignTokensOperation> alignmentOperation = new();
+            List<AnchorIndentationOperation> anchorIndentationOperations = new();
 
-            var nodeIterator = pooledIterator.Object;
-            using (Logger.LogBlock(FunctionId.Formatting_IterateNodes, cancellationToken))
-            {
-                const int magicLengthToNodesRatio = 5;
-                nodeIterator.Capacity = Math.Max(nodeIterator.Capacity, Math.Max(this.SpanToFormat.Length / magicLengthToNodesRatio, 4));
-
-                foreach (var node in _commonRoot.DescendantNodesAndSelf(this.SpanToFormat))
-                {
-                    cancellationToken.ThrowIfCancellationRequested();
-                    nodeIterator.Add(node);
-                }
-            }
-
-            // iterate through each operation using index to not create any unnecessary object
-            cancellationToken.ThrowIfCancellationRequested();
-            List<IndentBlockOperation> indentBlockOperation;
-            using (Logger.LogBlock(FunctionId.Formatting_CollectIndentBlock, cancellationToken))
-            {
-                indentBlockOperation = AddOperations<IndentBlockOperation>(nodeIterator, _formattingRules.AddIndentBlockOperations, cancellationToken);
-            }
-
-            cancellationToken.ThrowIfCancellationRequested();
-            List<SuppressOperation> suppressOperation;
-            using (Logger.LogBlock(FunctionId.Formatting_CollectSuppressOperation, cancellationToken))
-            {
-                suppressOperation = AddOperations<SuppressOperation>(nodeIterator, _formattingRules.AddSuppressOperations, cancellationToken);
-            }
+            List<IndentBlockOperation> indentBlockOperationScratch = new();
+            List<SuppressOperation> suppressOperationScratch = new();
+            List<AlignTokensOperation> alignmentOperationScratch = new();
+            List<AnchorIndentationOperation> anchorIndentationOperationsScratch = new();
 
-            cancellationToken.ThrowIfCancellationRequested();
-            List<AlignTokensOperation> alignmentOperation;
-            using (Logger.LogBlock(FunctionId.Formatting_CollectAlignOperation, cancellationToken))
+            // iterating tree is very expensive. only do it once.
+            foreach (var node in _commonRoot.DescendantNodesAndSelf(this.SpanToFormat))
             {
-                var operations = AddOperations<AlignTokensOperation>(nodeIterator, _formattingRules.AddAlignTokensOperations, cancellationToken);
-
-                // make sure we order align operation from left to right
-                operations.Sort((o1, o2) => o1.BaseToken.Span.CompareTo(o2.BaseToken.Span));
+                cancellationToken.ThrowIfCancellationRequested();
 
-                alignmentOperation = operations;
+                AddOperations(indentBlockOperation, indentBlockOperationScratch, node, _formattingRules.AddIndentBlockOperations);
```

Comment:
i'd prefer teh scratch buffers be cleared here if possible.

Reply:
➡️ Keeping since the current approach handles each one uniformly